### ISSUE TITLE
[ENH]: Allow `plt.sca(ax)` and/or ax to act as a context manager

### ISSUE THREAD

**Issue by ianhi** [matplotlib/matplotlib#20796]
### Problem

As a plot grows from a simple thing to a more complex thing and I add more and more subplots I often find myself with some pyplot code, or having to keep on writing things like
```python

axs[0,0]....
axs[0,0]....
axs[0,0]....
axs[0,0]....

axs[0,1]....
axs[0,1]....
axs[0,1]....
axs[0,1]....
```

This often happens when I'm not really ready to push things into functions yet so I end up typing all the above out many times, and if I need to change something it's a bit of a hassle.

### Proposed solution

The ability to do one or both of the following would be nice:

```python
with plt.sca(axs[0,1]):
   ...
```
or
```python
with axs[0,1] as ax:
   ...
```

### Additional context and prior art

The `sca` could be handled similarly to how `ion` is handled:
https://github.com/matplotlib/matplotlib/blob/ec126b0175c657f0adb5ca635eeabad7c142bb0b/lib/matplotlib/pyplot.py#L412-L435

and for the second option `ax` would just need to define these two methods:

```python
    def __enter__(self):
        return self
    def __exit__(self, exc_type, exc_value, traceback):
        pass
```

**Comment by ianhi** [comment]
This is also pretty easy to achieve on your own by doing this:
```python
from matplotlib.axes import Axes
def fake_enter(self):
    return self
def fake_exit(self, exc_type, exc_value, traceback):
    pass
Axes.__enter__ = fake_enter
Axes.__exit__= fake_exit

```

**Comment by jklymak** [comment]
What is wrong with `ax = axs[0, 1]`?

**Comment by ianhi** [comment]
I've done that as well and it does solve most of the issue, but I like visual distinction of indenting. Also I admit I originally wrote this about `plt.sca` which is slightly different and then got distrcted by wanting to do `with ax`

**Comment by timhoffm** [comment]
```
with plt.sca(axs[0,1]):
   ...
```
is reasonable.

```
with axs[0,1] as ax:
   ...
```
is too cute. It looks nice, but IMHO misuses the concept of a context manager for simple variable naming. On a side note, I always wanted to try writing a real local variable context `with localvars(ax=axs[0,1])`, so that `ax` can be used inside the context, but a `ax` outside would not be affected.

**Comment by tacaswell** [comment]
@anntzer has also had thoughts about this.

The idea of a current axes context has been kicking around for a probably pushing 10 years now (based on where I was sitting for the first round of these ideas!).

**Comment by jklymak** [comment]
I don't think we want to add functionality to the state-based interface, though, which would be the only reason to have a context manager do anything with `pyplot.sca`.

**Comment by ianhi** [comment]
I forgot the other major motivator for `plt.sca` (which is what I was actually doing) which is that this helpful when using 3rd party packages.

I know the advice is that 3rd party  functions should take `ax` as their first argument, but as a user of those functions I very much dislike having to add that to every call. So until there is a way to register new plotting functions to all ax objects (outside of a projection) I suspect that some 3rd party libs will continue to be at least somewhat `pyplot-like` accepting ax as an optional arg and calling `gca`. `mpl-interactions` does this everywhere and both `animatplot`, `mplcyberpunk` and  `seaborn` also seem to do this.

A consequence of this `psuedo-pyplot` pattern is that when I want to make a complex interactive visualization with subplots I end up having to add an
`ax=axs[...]` to every function call which can really only be made easier by setting the current axis hence my desire.. 

> I don't think we want to add functionality to the state-based interface, though, which would be the only reason to have a context manager do anything with pyplot.sca.

:( but understandable I guess. Depending on how this shakes out I can always include something like this myself in the fake pyplot that I've created. 


> On a side note, I always wanted to try writing a real local variable context with localvars(ax=axs[0,1]), so that ax can be used inside the context, but a ax outside would not be affected.

I also think that this would be rad - but somehow it also feels like it's maybe a "wrong" thing to do?

**Comment by jklymak** [comment]
I don't understand how 

```python
with plt.sca(axs[0]):
  iplt.plot(E, single_stark, controls=ctrls, ylim='auto')
  plt.ylabel('ΔI/I')
```
is any better than:

```python
plt.sca(axs[0])
iplt.plot(E, single_stark, controls=ctrls, ylim='auto')
plt.ylabel('ΔI/I')
```

It is more typing, and it doesn't do anything different.

**Comment by tacaswell** [comment]
The context manager shows you which axes is the "current axes" for than block of code.   Looking at

```python
with plt.sca(axs[0]):
   iplt.plot(E, single_stark, controls=ctrls, ylim='auto')
   plt.ylabel('ΔI/I')
with plt.sca(axs[0]):
   iplt.plot(E, single_stark, controls=ctrls, ylim='auto')
   plt.ylabel('ΔI/I')
with plt.sca(axs[0]):
   iplt.plot(E, single_stark, controls=ctrls, ylim='auto')
   plt.ylabel('ΔI/I')
with plt.sca(axs[0]):
   iplt.plot(E, single_stark, controls=ctrls, ylim='auto')
   plt.ylabel('ΔI/I')
```

vs

```python
plt.sca(axs[0])
iplt.plot(E, single_stark, controls=ctrls, ylim='auto')
plt.ylabel('ΔI/I')
plt.sca(axs[0])
iplt.plot(E, single_stark, controls=ctrls, ylim='auto')
plt.ylabel('ΔI/I')
plt.sca(axs[0])
iplt.plot(E, single_stark, controls=ctrls, ylim='auto')
plt.ylabel('ΔI/I')
plt.sca(axs[0])
iplt.plot(E, single_stark, controls=ctrls, ylim='auto')
plt.ylabel('ΔI/I')
```

**Comment by ianhi** [comment]
> It is more typing, and it doesn't do anything different.

Note that it is (could be) slightly different as I imagine that as context manager it will restore the original current axis. The equivalent would be:

```python
orig_ax = plt.gca()
plt.sca(axs[0])
# ...
plt.sca(orig_ax)
```

**Comment by timhoffm** [comment]
We should carefully consider naming. `gca()` / `sca()` is short but quite opaque. Further I assume that `sca()` is quite rarely used, so introducing an new name for the context manager would be ok. An alternative could be:
```
with plt.current_axes(axs[0]):
    ...
```

**Comment by tacaswell** [comment]
Context manager is still an interesting idea, but I am not sure we should keep an open issue for it.

**Comment by timhoffm** [comment]
Reconsidering: IMHO we don't need this. This is rightfully closed as "not planned".

While the proposal is technical technical sound, we don't gain much but increase complexity.

Arguments:

- **Using mulitple Axes on pyplot is generally not recommended**:

  From https://matplotlib.org/devdocs/users/explain/quick_start.html#the-explicit-and-the-implicit-interfaces:

  > In general, we suggest using the OO style, particularly for complicated plots, and functions and scripts that are intended to be reused as part of a larger project. However, the pyplot style can be very convenient for quick interactive work.

  Overall, when using more than one Axes, people should use the OO interface. Not that pyplot is prohibited, but we don't have to create extra API for better multi-Axes handling in pyplot.

- **If a reference to the Axes already exists, why not simply use its methods?**

  For `with plt.sca(ax)` you aleady have the variable. Why do
  ```python
  with plt.sca(ax):
      plt.plot(...)
      plt.title(...)
  ```
  when you can instead simply do
  ```python
  ax.plot(...)
  ax.set_title(...)
  ```

- **Context does not help with interactive incremental operation**
  The context cannot stretch across multiple REPL commands. I.e. you have to write everything into one REPL command. So likely not helpful for interactive use, and even more difficult to write because of the multi-line and indentation.

- **The context does not gain much compared to a fixed sca().**
  The only advantage of the context compared to a simple `plt.sca(ax)` is that the previous current axes is restored. However, cases where you *temporarily* work on another Axes and come back to a previous Axes should be rare. Typically you'll configure one Axes completely and then move to the next. 

- **Yet another way to doc things**
  We would need to teach the context manager, and it's yet another way to handle multiple Axes on top of the existing "permanently set pyplot global state" and "hold references to all Axes and call their methods".

**Comment by tacaswell** [comment]
That is a pretty compelling argument against it.  Best counter arguement I have is to make dealing with ~legacy~  heritage code that uses pyplot internally and can not / do not want to retrofit that code so

```python
with plt.sca(ax):
    some_function_that_uses_plt(...)
```

This might be more compelling now that we have the ability for pyplot to adopt figure so this could be a tool to help smooth the transition from pyplot using to fully non-pyplot code.

**Comment by timhoffm** [comment]
That's indeed the only argument in favor of the context manager.

Still my fourth point (The context does not gain much compared to a fixed sca()) does apply.

I think it's not worth the effort, but I wouldn't block a contextmanager if others want to implement it.
